clang/lib/Basic/Targets/PPC.cpp

```diff
@@ -807,7 +807,7 @@ ArrayRef<TargetInfo::GCCRegAlias> PPCTargetInfo::getGCCRegAliases() const {
 // PPC ELFABIv2 DWARF Definitoin "Table 2.26. Mappings of Common Registers".
 // vs0 ~ vs31 is mapping to 32 - 63,
 // vs32 ~ vs63 is mapping to 77 - 108.
```

Comment:
I am not sure it helps to appeal to the ELFABIv2 DWARF definition if we are no longer restraining this data to be used only under ELFABIv2. Instead, the usage here can be explained as the index numbers for the related FPRs and VRs in `GCCRegNames` above.

Reply:
This is what I get from DWARF spec:

Note that the register number represents a DWARF specific mapping of numbers onto the actual registers of a given architecture. The mapping should be chosen to gain optimal density and should be shared by all users of a given architecture. It is recommended that this mapping be defined by the ABI authoring committee for each architecture.

To me, the DWARF number should be architecture specific, i.e., although the ABIs are different, all these ABIs on same architecture should use same DWARF register mapping, because the hardware registers are bound to architecture?

Reply:
> To me, the DWARF number should be architecture specific, i.e., although the ABIs are different, all these ABIs on same architecture should use same DWARF register mapping, because the hardware registers are bound to architecture?

While that makes sense, I am not seeing elsewhere in the code comments that the DWARF numbers are being used. The code _does_ support (in `getNormalizedGCCRegisterName`) that these numbers are used to index into the `GCCRegNames` array.

Reply:
FP and VMX registers should have different DWARF number with VSX registers although on PPC they are physically overlapped.

Let me check the usage for the DWARF numbers...

Reply:
@hubert-reinterpretcast I think you are right. There is a legacy issue in the function `getNormalizedGCCRegisterName()`. Once `ReturnCanonical` is true for that function, current clang(without this patch) will use name f0 ~ f31 and v0 ~ v31 for vs0 ~ vs63 on ELFABIV2. And this patch extends this "issue" to AIX.

The `ReturnCanonical` was introduced in https://reviews.llvm.org/D15075 which was targeted for X86. Seems for X86, physically overlapping registers have canonical names, for example X86 ([eax|rax|al|ah] -> ax). But targets like PPC and SystemZ do not have such canonical register names, for example, PPC (vs0 -> f0), SystemZ(v0 -> f0) which are not right.

I prefer to solve this in another patch with solution: for targets that have no canonical name for physically overlapping registers, `ReturnCanonical` should always be false.

What do you think?

Reply:
> I prefer to solve this in another patch with solution: for targets that have no canonical name for physically overlapping registers, `ReturnCanonical` should always be false.
> 
> What do you think?

I think the usage of `ReturnCanonical` is at least partially intended (and applies to the situation with vs0 -> f0). In particular, it is intended to diagnose (as GCC does):
```
void f(void) {
  register float f __asm__("fr1");
  __asm__ __volatile__ (
    "fmul %0,%0,%0"
    :
    : "f"(f)
    : "vs1"
  );
}
```

That diagnostic does not seem to operate for Clang (at least for PPC) even with a simpler case where the clobber is `fr1` though.

I think we can leave that for later. For now, with the current patch, I think a comment to say that these numbers are used for indexing into the `GCCRegNames` array would be correct. If you believe that is a problem, then an additional FIXME comment can be added.

Reply:
Done.